### PR TITLE
assimp: update to 5.2.1

### PR DESCRIPTION
**Pull request by hgdagon**
Tested on MinGW64, hoping the tests will do the rest. Also planning to suggest these changes to assimp repo, so by the naext version most of these changes should be unneeded.

### CODE REVIEW

File: mingw-w64-assimp/msys2-patches.patch [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
++++ assimp-5.2.1/code/CMakeLists.txt	2022-02-16 23:54:18.000000000 -0800
+@@ -1293,12 +1293,16 @@
+          "${HEADER_PATH}/Compiler"
+          assimp.framework/Headers/Compiler
+       COMMENT "Copying public ./Compiler/ header files to framework bundle's Headers/Compiler/")
+   ENDIF()
+ ENDIF()
+ 
++IF ( MSYS  AND NOT BUILD_SHARED_LIBS)
++  TARGET_LINK_LIBRARIES ( assimp -static-libgcc -static-libstdc++ -Wl,-Bstatic -lstdc++ -lwinpthread )
```
Comment (Biswa96) [inline mingw-w64-assimp/msys2-patches.patch:11]: `-static-libstdc++` option will show warnings in clang: warning: argument unused during compilation: '-static-libstdc++'

Comment (hgdagon) [reply mingw-w64-assimp/msys2-patches.patch:11]: Removed that section entirely. That was for the future anyway. We're only building shared so far.

File: mingw-w64-assimp/msys2-patches.patch [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+       # Multibyte character set is deprecated since at least MSVC2015 (possibly earlier)
+       ADD_DEFINITIONS( -DUNICODE -D_UNICODE )
+@@ -166,12 +162,15 @@
+   MESSAGE(STATUS "Framework bundle building enabled")
+ ENDIF()
+ 
+ IF(NOT BUILD_SHARED_LIBS)
+   MESSAGE(STATUS "Shared libraries disabled")
+   SET(LINK_SEARCH_START_STATIC TRUE)
++  IF(MSYS)
```
[the remark this answers is not in the document]
Comment (Biswa96) [reply mingw-w64-assimp/msys2-patches.patch:46]: I am wondering if this required at all. Is not the extension automatically set by cmake? What does happen in Linux system? We may follow that condition.

[the remark this answers is not in the document]
Comment (hgdagon) [reply mingw-w64-assimp/msys2-patches.patch:46]: Don't have Linux machine to test, atm. But on windows/msys2, CMake detects shared libs by default, and we want a full static build. This change was also for assimp itself, not for this PKGBUILD, so I removed this from the patch as well. So far, it seems like building both static and shared is not an option, so I'll stick to shared build only for now.

**Comment by hgdagon** [comment]
```c++
#if defined _WIN32 && (!defined __GNUC__ || __MSVCRT_VERSION__ >= 0x0601)
```

This check evaluates to false on Clang32, resulting in a build error.

```c++
#if defined _WIN32 && !defined __GNUC__
```

This, however evaluates to true. There is no `__MSVCRT_VERSION__` definition anywhere, though...

Maybe this is an issue with MinGW and should be reported there?

**Comment by MehdiChinoune** [comment]
The last time we tried to update assimp, building both qt6-3d and qt5-3d failed.

**Comment by jeremyd2019** [comment]
It looks like the DLL name changed, so at a minimum these packages would need a rebuild:
* mingw-w64-magnum-plugins
* mingw-w64-qt5-pkgs
* mingw-w64-qt5-3d
* mingw-w64-qt6-3d

(I didn't expect the DLL name had changed, and was pleasantly surprised that it successfully tested a different case in my new script)

**Comment by jeremyd2019** [comment]
```
2022-02-17T22:29:58.1838543Z In file included from assimpimporter.cpp:37:
2022-02-17T22:29:58.3739375Z D:/M/msys64/clang64/include/assimp/pbrmaterial.h:51:5: warning: pbrmaterial.h is deprecated. Please update to PBR materials in materials.h and glTF-specific items in GltfMaterial.h [-W#warnings]
2022-02-17T22:29:58.5678841Z #   warning pbrmaterial.h is deprecated. Please update to PBR materials in materials.h and glTF-specific items in GltfMaterial.h
2022-02-17T22:29:58.7598078Z     ^
2022-02-17T22:29:59.1479258Z assimpimporter.cpp:609:10: warning: variable 'needsUV2Data' set but not used [-Wunused-but-set-variable]
2022-02-17T22:29:59.2296939Z     bool needsUV2Data = false;
2022-02-17T22:29:59.4097169Z          ^
2022-02-17T22:29:59.4932004Z assimpimporter.cpp:608:10: warning: variable 'needsUV1Data' set but not used [-Wunused-but-set-variable]
2022-02-17T22:29:59.6096664Z     bool needsUV1Data = false;
2022-02-17T22:29:59.6836653Z          ^
2022-02-17T22:29:59.7638947Z assimpimporter.cpp:1086:32: error: use of undeclared identifier 'AI_MATKEY_GLTF_PBRSPECULARGLOSSINESS'
2022-02-17T22:29:59.7812593Z         result = material->Get(AI_MATKEY_GLTF_PBRSPECULARGLOSSINESS, hasSpecularGlossiness);
2022-02-17T22:29:59.7860349Z                                ^
```

**Comment by hgdagon** [comment]
@jeremyd2019 Do you think yo can edit the 3d and quick3d packages for both qt5 and qt6 to use their own assimp? I couldn't figure it out, but since they both come with their own, it's not really worth it to keep this package back. That said, I don't really know how much those packages need to be reworked to support the latest assimp. In case of `AI_MATKEY_GLTF_PBRSPECULARGLOSSINESS`, it used to be in  `pbrmaterial.h`, and now it's in `GltfMaterial.h`, but I doubt it's the only incompatibility (Although it could be, since if I understand correctly, qt uses assimp for Gltf only).

**Comment by MehdiChinoune** [comment]
I will push a PR to upgrade assimp with a patch that allows qt5-3d, qt5-quick3d, qt6-3d and qt6-quick3d to be built successfully once #10791 get merged (built and uploaded to staging repo)

File: mingw-w64-assimp/msys2-patches.patch [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+--- assimp-5.2.1/cmake-modules/FindDirectX.cmake.orig	2022-02-07 01:02:13.000000000 -0800
++++ assimp-5.2.1/cmake-modules/FindDirectX.cmake	2022-02-15 23:59:25.000000000 -0800
+@@ -34,12 +34,13 @@
+     "C:/apps_x86/Microsoft DirectX SDK*"
+     "C:/Program Files (x86)/Microsoft DirectX SDK*"
+     "C:/apps/Microsoft DirectX SDK*"
+     "C:/Program Files/Microsoft DirectX SDK*"
+     "C:/Program Files (x86)/Windows Kits/8.1"
+     "$ENV{ProgramFiles}/Microsoft DirectX SDK*"
++    "$ENV{MSYSTEM_PREFIX}/$ENV{MSYSTEM_CHOST}"
```
Comment (jeremyd2019) [inline mingw-w64-assimp/msys2-patches.patch:100]: this path won't be right anymore, it should just be ""$ENV{MSYSTEM_PREFIX}"
```suggestion
+    "$ENV{MSYSTEM_PREFIX}"
```

File: mingw-w64-assimp/msys2-patches.patch [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+ template <>
+@@ -146,13 +146,13 @@
+         //  - For binary streams, it is not technically well-defined
+         //  - For text files the results are meaningless
+         // That's why we use the safer variant fstat here.
+         //
+         // See here for details:
+         // https://www.securecoding.cert.org/confluence/display/seccode/FIO19-C.+Do+not+use+fseek()+and+ftell()+to+compute+the+size+of+a+regular+file
+-#if defined _WIN32 && (!defined __GNUC__ || __MSVCRT_VERSION__ >= 0x0601)
++#if defined _WIN64 && (!defined __GNUC__ || __MSVCRT_VERSION__ >= 0x0601)
```
Comment (jeremyd2019) [inline mingw-w64-assimp/msys2-patches.patch:84]: I think this blob would be fine without being patched.

Comment (hgdagon) [reply mingw-w64-assimp/msys2-patches.patch:84]: It's not. In Clang32 , this evaluates to true resulting in a build error.

**Comment by lazka** [comment]
Can this be closed because of #10805, or is there anything still missing?

**Comment by hgdagon** [comment]
> Can this be closed because of #10805, or is there anything still missing?

Nope.

**Pull request closed without merging**